platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/AssignImmutableWorkspaceStep.java

```diff
@@ -104,19 +138,24 @@ private Optional<WorkspaceResult> loadImmutableWorkspaceIfExists(UnitOfWork work
         }
     }
 
-    private WorkspaceResult loadImmutableWorkspace(UnitOfWork work, File immutableLocation) {
+    private Optional<WorkspaceResult> loadImmutableWorkspaceIfConsistent(UnitOfWork work, ImmutableWorkspace workspace) {
+        File immutableLocation = workspace.getImmutableLocation();
         ImmutableSortedMap<String, FileSystemSnapshot> outputSnapshots = outputSnapshotter.snapshotOutputs(work, immutableLocation);
 
         // Verify output hashes
         ImmutableListMultimap<String, HashCode> outputHashes = calculateOutputHashes(outputSnapshots);
         ImmutableWorkspaceMetadata metadata = workspaceMetadataStore.loadWorkspaceMetadata(immutableLocation);
         if (!metadata.getOutputPropertyHashes().equals(outputHashes)) {
-            throw new IllegalStateException(
-                "Immutable workspace contents have been modified: " + immutableLocation.getAbsolutePath() + ". " +
-                    "These workspace directories are not supposed to be modified once they are created. " +
-                    "Deleting the directory in question can allow the content to be recreated.");
+            boolean moveSuccessful = workspace.withTemporaryWorkspace(temporaryWorkspace -> moveInconsistentImmutableWorkspaceToTemporaryLocation(immutableLocation, temporaryWorkspace, outputSnapshots));
+            if (moveSuccessful) {
+                return Optional.empty();
+            }
```

Comment:
💭 This is the heart of the change: if we find an inconsistent workspace, we don't throw, but instead try to move it out of the way. If we managed to move it out of the way, we act as if no existing workspace was found (`Optional.empty()`). If we fail to move, we fall back to trying to reuse what we found.